sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -41,6 +42,11 @@ case class Mode(
     this(child, 0, 0, Some(reverse))
   }
 
+  final lazy val collationId: Int = child.dataType match {
```

Comment:
if we use `asInstanceOf[StringType]` we will get an exception whenever the child is a numerical type or any non-string type.

It would be ideal in the case of a non-string type if collationId could be -1 or something, but then CollationFactory.fetchCollation(collationId) would throw an exception. But it really is not precise to be relying on `!CollationFactory.fetchCollation(CollationFactory.UTF8_BINARY_COLLATION_ID).supportsBinaryEquality` when it is a non-string-type to continue with the non-collation logic.